src/Raven.Server/Documents/ShardedHandlers/ShardedQueriesHandler.cs

```diff
@@ -0,0 +1,843 @@
+
+            // User should also be able to define a query parameter ("Sharding.Context") which is an array
+            // that contains the ids whose shards the query should be limited to. Advanced: Optimization if user
+            // want to run a query and knows what shards it is on. Such as:
+            // from Orders where State = $state and User = $user where all the orders are on the same share as the user
+
+            // - TODO: Sharding - etag handling!
+            public ShardedQueryProcessor(TransactionOperationContext context,ShardedQueriesHandler parent,IndexQueryServerSide query)
+            {
+                _context = context;
+                _result = new ShardedQueryResult();
+                _parent = parent;
+                _query = query;
+                _commands = new Dictionary<int, ShardedQueryCommand>();
+                if (_query.QueryParameters != null && _query.QueryParameters.TryGet("Sharding.Context", out BlittableJsonReaderArray filter))
+                {
+                    _filteredShardIndexes = new HashSet<int>();
+                    foreach (object item in filter)
+                    {
+                        _filteredShardIndexes.Add(_parent.ShardedContext.GetShardIndex(_context, item.ToString()));
+                    }
+                }
+                else
+                {
+                    _filteredShardIndexes = null;
+                }
+            }
+
+            public long ResultsEtag => _result.ResultEtag;
+
+            public void Initialize()
+            {
+                if (_query.Metadata.IsGraph)
+                {
+                    // * Graph queries - not supported for sharding
+                    throw new NotSupportedException("Graph queries aren't supported for sharding");
+                }
+
+                // now we have the index query, we need to process that and decide how to handle this.
+                // There are a few different modes to handle:
+                var queryTemplate = _query.ToJson(_context);
+                if (_query.Metadata.IsCollectionQuery)
+                {
+                    // * For collection queries that specify ids, we can turn that into a set of loads that 
+                    //   will hit the known servers
+
+                    (List<string> ids, string _) = _query.ExtractIdsFromQuery(_parent.ServerStore, _parent.ShardedContext.DatabaseName);
+                    if (ids != null)
+                    {
+                        GenerateLoadByIdQueries(ids, _commands);
+                        return;
+                    }
+                }
+                
+                // * For paging queries, we modify the limits on the query to include all the results from all
+                //   shards if there is an offset. But if there isn't an offset, we can just get the limit from
+                //   each node and then merge them
+                RewriteQueryForPaging(ref queryTemplate);
```

Comment:
if we have `Take` for a MapReduce index we need to request the result with order by.
Otherwise we might have partial results

Reply:
Added an issue